File: app/src/main/java/me/kr328/nevo/decorators/smscaptcha/CaptchaDecoratorService.java
package me.kr328.nevo.decorators.smscaptcha;

import android.app.KeyguardManager;
import android.app.Notification;
import android.content.BroadcastReceiver;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;
import android.widget.Toast;

import com.oasisfeng.nevo.sdk.MutableNotification;
import com.oasisfeng.nevo.sdk.MutableStatusBarNotification;

import java.util.Arrays;
import java.util.Objects;
import java.util.TreeSet;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import me.kr328.nevo.decorators.smscaptcha.utils.CaptchaUtils;
import me.kr328.nevo.decorators.smscaptcha.utils.NotificationUtils;
import me.kr328.nevo.decorators.smscaptcha.utils.PatternUtils;

public class CaptchaDecoratorService extends BaseSmsDecoratorService {
    public static final String TAG = CaptchaDecoratorService.class.getSimpleName();

    public static final String NOTIFICATION_EXTRA_RECAST = Global.PREFIX_NOTIFICATION_EXTRA + ".captcha.notification.recast";

    private Settings mSettings;
    private CaptchaUtils mCaptchaUtils;

    private TreeSet<String> mAppliedKeys = new TreeSet<>();

    private BroadcastReceiver mKeyguardReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Bundle fillBundle = new Bundle();
            fillBundle.putBoolean(NOTIFICATION_EXTRA_RECAST, true);

            recastAllNotifications(fillBundle);
        }
    };

    private BroadcastReceiver mCancelNotificationReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String key = intent.getStringExtra(Global.INTENT_NOTIFICATION_KEY);

            cancelNotification(key);
        }
    };

    @Override
    protected void apply(MutableStatusBarNotification evolving) {
        MutableNotification notification = evolving.getNotification();
        Bundle extras = notification.extras;
        boolean recast = extras.getBoolean(NOTIFICATION_EXTRA_RECAST, false);
        NotificationUtils.Messages messages = NotificationUtils.parseMessages(notification);
        String[] captchas = mCaptchaUtils.findSmsCaptchas(messages.texts);

        Log.i(TAG, "apply begin");

        Stream.of(messages).forEach((msg) -> Stream.of(msg.texts).forEach((c) -> Log.i(TAG, "Message " + c)));

        if (captchas.length == 0) {
            Log.i(TAG, "Captcha not found.");
            return;
        }

        if (mSettings.isCaptchaHideOnLocked()) {
            KeyguardManager keyguardManager = (KeyguardManager) getSystemService(Context.KEYGUARD_SERVICE);
            if (keyguardManager != null && keyguardManager.isKeyguardLocked())
                applyKeyguardLocked(notification, evolving.getKey(), messages, captchas);
            else
                applyKeyguardUnlocked(notification, evolving.getKey(), messages, captchas);
            notification.visibility = Notification.VISIBILITY_PUBLIC;
        } else {
            applyKeyguardLocked(notification, evolving.getKey(), messages, captchas);
        }

        notification.flags |= Notification.FLAG_ONLY_ALERT_ONCE;

        mAppliedKeys.add(evolving.getKey());

        if (!recast)
            sendBroadcast(new Intent(Global.INTENT_CAPTCHA_NOTIFICATION_SHOW).
                    putExtra(Global.INTENT_NOTIFICATION_KEY, evolving.getKey()).
                    putExtra(Global.INTENT_NOTIFICATION_CAPTCHA, captchas[0]));

        Log.i(TAG, "Applied " + evolving.getKey());
    }

    private void applyKeyguardLocked(Notification notification, String key, NotificationUtils.Messages messages, String[] captchas) {
        Notification.Action[] actions = new Notification.Action[]{
                createNonIconAction(key, getString(R.string.captcha_service_notification_locked_action_copy_code), new CaptchaMessage(messages, captchas[0]))
        };

        NotificationUtils.replaceMessages(notification, text -> CaptchaUtils.replaceCaptchaWithChar(text, captchas, '*'));

        if (mSettings.isCaptchaOverrideDefaultAction())
            replaceActions(notification, key, actions);
        else
            appendActions(notification, key, actions);
    }

    private void applyKeyguardUnlocked(Notification notification, String key, NotificationUtils.Messages messages, String[] captchas) {
        Notification.Action[] actions = Arrays.stream(captchas).
                map(captcha -> createNonIconAction(key, getString(R.string.captcha_service_notification_unlocked_action_copy_code_format, captcha), new CaptchaMessage(messages, captcha))).
                toArray(Notification.Action[]::new);

        if (mSettings.isCaptchaOverrideDefaultAction())
            replaceActions(notification, key, actions);
        else
            appendActions(notification, key, actions);
    }

    private void loadSettings() {
        mSettings = Settings.fromApplication(getApplication());
    }

    private void initCaptchaUtils() {
        mCaptchaUtils = new CaptchaUtils(mSettings.isCaptchaUseDefaultPattern(),
                PatternUtils.compilePattern(mSettings.getCaptchaIdentifyPattern(), "", Pattern.CASE_INSENSITIVE),
                PatternUtils.compilePattern(mSettings.getCaptchaParsePattern(), "", Pattern.CASE_INSENSITIVE));

        Log.d(TAG, "CaptchaUtils " + mSettings.isCaptchaUseDefaultPattern() + " " + mSettings.getCaptchaIdentifyPattern() + " " + mSettings.getCaptchaParsePattern());
    }

    private void recastAllNotifications(Bundle fillInExtras) {
        for (String key : mAppliedKeys)
            recastNotification(key, fillInExtras);
        mAppliedKeys.clear();
    }

    private void copyCaptcha(String captcha, NotificationUtils.Messages messages) {
        ((ClipboardManager) Objects.requireNonNull(getSystemService(Context.CLIPBOARD_SERVICE))).
                setPrimaryClip(ClipData.newPlainText("SmsCaptcha", captcha));

        Toast.makeText(this, getString(R.string.captcha_service_toast_copied_format, captcha), Toast.LENGTH_LONG).show();
    }

    @Override
    public void onActionClicked(String key, Parcelable cookies) {
        CaptchaMessage captchaMessage = (CaptchaMessage) cookies;

        if (captchaMessage == null)
            return;

        copyCaptcha(captchaMessage.captcha, captchaMessage.messages);
    }

    private void registerReceivers() {
        registerReceiver(mKeyguardReceiver, new IntentFilter() {{
            addAction(Intent.ACTION_USER_PRESENT);
            addAction(Intent.ACTION_SCREEN_OFF);
        }});
        registerReceiver(mCancelNotificationReceiver, new IntentFilter(Global.INTENT_CAPTCHA_NOTIFICATION_DO_CANCEL));
    }

    private void unregisterReceivers() {
        unregisterReceiver(mKeyguardReceiver);
        unregisterReceiver(mCancelNotificationReceiver);
    }

    @Override
    protected void onNotificationRemoved(String key, int reason) {
        super.onNotificationRemoved(key, reason);

        sendBroadcast(new Intent(Global.INTENT_CAPTCHA_NOTIFICATION_CANCEL).putExtra(Global.INTENT_NOTIFICATION_KEY, key));

        mAppliedKeys.remove(key);
    }

    @Override
    protected void onConnected() {
        super.onConnected();

        loadSettings();
        initCaptchaUtils();
        registerReceivers();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        unregisterReceivers();
    }

    static class CaptchaMessage implements Parcelable {
        public static final Creator<CaptchaMessage> CREATOR = new Creator<CaptchaMessage>() {
            @Override
            public CaptchaMessage createFromParcel(Parcel source) {
                NotificationUtils.Messages messages = new NotificationUtils.Messages();
                messages.texts = new String[source.readInt()];

                for (int i = 0; i < messages.texts.length; i++)
                    messages.texts[i] = source.readString();

                return new CaptchaMessage(messages, source.readString());
            }

            @Override
            public CaptchaMessage[] newArray(int size) {
                return new CaptchaMessage[size];
            }
        };
        NotificationUtils.Messages messages;
        String captcha;

        CaptchaMessage(NotificationUtils.Messages messages, String captcha) {
            this.messages = messages;
            this.captcha = captcha;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeInt(messages.texts.length);

            for (CharSequence s : messages.texts) dest.writeString(s.toString());

            dest.writeString(captcha);
        }
    }
}

File: app/src/main/java/me/kr328/nevo/decorators/smscaptcha/CaptchaDecoratorService.kt
package me.kr328.nevo.decorators.smscaptcha

import android.app.KeyguardManager
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.*
import android.os.*
import android.util.Log
import android.widget.Toast
import androidx.annotation.Keep
import com.oasisfeng.nevo.sdk.MutableStatusBarNotification
import me.kr328.nevo.decorators.smscaptcha.compat.NOTIFICATION_PRIORITY_HIGH_COMPAT
import me.kr328.nevo.decorators.smscaptcha.compat.priorityCompat
import me.kr328.nevo.decorators.smscaptcha.utils.CaptchaUtils
import me.kr328.nevo.decorators.smscaptcha.utils.NotificationUtils.parseMessages
import me.kr328.nevo.decorators.smscaptcha.utils.NotificationUtils.replaceMessages
import me.kr328.nevo.decorators.smscaptcha.utils.PackageUtils.hasPackageInstalled
import me.kr328.nevo.decorators.smscaptcha.utils.PatternUtils.compilePattern
import net.grandcentrix.tray.AppPreferences
import net.grandcentrix.tray.core.TrayItem
import java.lang.Exception
import java.util.*

class CaptchaDecoratorService : BaseSmsDecoratorService() {
    private lateinit var mSettings: Settings
    private lateinit var mCaptchaUtils: CaptchaUtils
    private val mAppliedMessages = mutableMapOf<String, List<String>>()
    private val mKeyguardReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            recastAllNotifications()
        }
    }

    override fun apply(evolving: MutableStatusBarNotification): Boolean {
        try {
            val notification = evolving.notification
            val extras = notification.extras
            val messages = (mAppliedMessages[evolving.key] ?: emptyList()) + parseMessages(notification)
            val captchas = mCaptchaUtils.extractSmsCaptchas(messages, mSettings.isCaptchaUseDefaultPattern)

            if (captchas.isEmpty()) return false

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                notification.channelId = NOTIFICATION_CHANNEL_CAPTCHA
            } else {
                notification.priorityCompat = NOTIFICATION_PRIORITY_HIGH_COMPAT
            }

            val keyguardManager = getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager

            if (mSettings.isCaptchaHideOnLocked && keyguardManager.isKeyguardLocked)
                applyKeyguardLocked(notification, evolving.key, messages, captchas)
            else
                applyKeyguardUnlocked(notification, evolving.key, messages, captchas)

            notification.flags = notification.flags or Notification.FLAG_ONLY_ALERT_ONCE
            notification.visibility = Notification.VISIBILITY_PUBLIC
            extras.putBoolean(Constants.NOTIFICATION_EXTRA_APPLIED, true)
            mAppliedMessages[evolving.key] = messages

            Log.i(Constants.TAG, "Applied " + evolving.key)

            return true
        }
        catch (e: Exception) {
            Log.w(Constants.TAG, "Apply failure", e)
            return false
        }
    }

    private fun applyKeyguardLocked(notification: Notification, key: String, messages: List<String>, captchas: List<String>) {
        val actions = arrayOf(
                createNonIconAction(key, getString(R.string.captcha_service_notification_locked_action_copy_code), CaptchaMessage(messages, captchas[0]))
        )
        replaceMessages(notification) { text: String -> mCaptchaUtils.replaceCaptchaWithChar(text, captchas) }

        if (mSettings.isCaptchaOverrideDefaultAction)
            replaceActions(notification, actions)
        else
            appendActions(notification, key, actions)
    }

    private fun applyKeyguardUnlocked(notification: Notification, key: String, messages: List<String>, captchas: List<String>) {
        val actions = captchas.map {
            createNonIconAction(key, getString(R.string.captcha_service_notification_unlocked_action_copy_code_format, it), CaptchaMessage(messages, it))
        }.toTypedArray()

        if (mSettings.isCaptchaOverrideDefaultAction)
            replaceActions(notification, actions)
        else
            appendActions(notification, key, actions)
    }

    private fun loadSettings() {
        if (Objects.requireNonNull(getSystemService(UserManager::class.java)).isUserUnlocked) {
            val mAppPreferences = AppPreferences(this)
            mSettings = Settings().applyFromTrayPreference(mAppPreferences)
            mAppPreferences.registerOnTrayPreferenceChangeListener(this::onSettingsChanged)
        } else {
            mSettings = Settings()
        }
    }

    private fun createNotificationChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(NOTIFICATION_CHANNEL_CAPTCHA,
                    getString(R.string.captcha_service_notification_channel_name),
                    NotificationManager.IMPORTANCE_HIGH)

            for (packageName in DEFAULT_TARGET_PACKAGES)
                if (hasPackageInstalled(this, packageName)) createNotificationChannels(packageName, Process.myUserHandle(), listOf(channel))
        }
    }

    private fun initCaptchaUtils() {
        mCaptchaUtils = CaptchaUtils(this,
                compilePattern(mSettings.captchaIdentifyPattern, "", setOf(RegexOption.IGNORE_CASE)),
                compilePattern(mSettings.captchaParsePattern, "", setOf(RegexOption.IGNORE_CASE)))
    }

    private fun recastAllNotifications() {
        for (entry in mAppliedMessages) recastNotification(entry.key, null)
        mAppliedMessages.clear()
    }

    private fun copyCaptcha(captcha: String) {
        (Objects.requireNonNull(getSystemService(Context.CLIPBOARD_SERVICE)) as ClipboardManager)
                .setPrimaryClip(ClipData.newPlainText("SmsCaptcha", captcha))
        Toast.makeText(this, getString(R.string.captcha_service_toast_copied_format, captcha), Toast.LENGTH_LONG).show()
    }

    override fun onUserUnlocked() {
        loadSettings()
    }

    override fun onActionClicked(key: String, cookies: Parcelable?) {
        val captchaMessage = cookies as CaptchaMessage? ?: return
        copyCaptcha(captchaMessage.captcha)
    }

    private fun registerReceivers() {
        registerReceiver(mKeyguardReceiver, object : IntentFilter() {
            init {
                addAction(Intent.ACTION_USER_PRESENT)
                addAction(Intent.ACTION_SCREEN_OFF)
            }
        })
    }

    private fun unregisterReceivers() {
        unregisterReceiver(mKeyguardReceiver)
    }

    override fun onNotificationRemoved(key: String, reason: Int): Boolean {
        super.onNotificationRemoved(key, reason)

        mAppliedMessages.remove(key)

        return false
    }

    override fun onConnected() {
        super.onConnected()

        loadSettings()
        initCaptchaUtils()
        createNotificationChannels()
        registerReceivers()
    }

    private fun onSettingsChanged(trayItems: Collection<TrayItem>) {
        for (item in trayItems) {
            when (item.key()) {
                Settings.SETTING_CAPTCHA_IDENTIFY_PATTERN -> {
                    mSettings.captchaIdentifyPattern = item.value().orEmpty()
                    initCaptchaUtils()
                }
                Settings.SETTING_CAPTCHA_PARSE_PATTERN -> {
                    mSettings.captchaParsePattern = item.value().orEmpty()
                    initCaptchaUtils()
                }
                Settings.SETTING_CAPTCHA_USE_DEFAULT_PATTERN -> {
                    mSettings.isCaptchaUseDefaultPattern = item.value()?.toBoolean() ?: true
                    initCaptchaUtils()
                }
                Settings.SETTING_CAPTCHA_HIDE_ON_LOCKED -> mSettings.isCaptchaHideOnLocked = item.value()?.toBoolean()
                        ?: false
                Settings.SETTING_CAPTCHA_OVERRIDE_DEFAULT_ACTION -> mSettings.isCaptchaOverrideDefaultAction = item.value()?.toBoolean()
                        ?: false
            }
            Log.i(Constants.TAG, "Settings Updated " + item.key() + "=" + item.value())
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceivers()
    }

    @Keep
    data class CaptchaMessage(val messages: List<String>, val captcha: String) : Parcelable {
        constructor(parcel: Parcel) : this(
                parcel.createStringArrayList() ?: throw NullPointerException(),
                parcel.readString() ?: throw NullPointerException())

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeStringList(messages)
            dest.writeString(captcha)
        }

        companion object CREATOR : Parcelable.Creator<CaptchaMessage> {
            override fun createFromParcel(parcel: Parcel): CaptchaMessage {
                return CaptchaMessage(parcel)
            }

            override fun newArray(size: Int): Array<CaptchaMessage?> {
                return arrayOfNulls(size)
            }
        }
    }

    companion object {
        val DEFAULT_TARGET_PACKAGES = listOf(
                "com.android.messaging",
                "com.google.android.apps.messaging",
                "com.android.mms",
                "com.sonyericsson.conversations",
                "com.moez.QKSMS"
        )
        const val NOTIFICATION_CHANNEL_CAPTCHA = "notification_channel_captcha"
        const val NOTIFICATION_EXTRA_RECAST = Constants.PREFIX_NOTIFICATION_EXTRA + ".captcha.notification.recast"
    }
}